src/searching/ternary_search.rs

```diff
@@ -1,91 +1,192 @@
+    let cmp_second_mid = item.cmp(&arr[second_mid]);
 
-    #[test]
-    fn returns_index_if_list_has_one_item() {
-        let index = ternary_search(&1, &[1], 0, 1);
-        assert_eq!(index, Some(0));
-    }
-
-    #[test]
-    fn returns_first_index() {
-        let index = ternary_search(&1, &[1, 2, 3], 0, 2);
-        assert_eq!(index, Some(0));
+    match (is_asc, cmp_first_mid, cmp_second_mid) {
+        // If the item matches either midpoint, it returns the index
+        (_, Ordering::Equal, _) => {
+            *left = first_mid;
+            return true;
+        }
+        (_, _, Ordering::Equal) => {
+            *left = second_mid;
+            return true;
+        }
+        // If the item is smaller than the element at first_mid (in ascending order)
+        // or greater than it (in descending order), it narrows the search to the first third.
+        (true, Ordering::Less, _) | (false, Ordering::Greater, _) => {
+            *right = first_mid.saturating_sub(1)
+        }
+        // If the item is greater than the element at second_mid (in ascending order)
+        // or smaller than it (in descending order), it narrows the search to the last third.
+        (true, _, Ordering::Greater) | (false, _, Ordering::Less) => *left = second_mid + 1,
+        // Otherwise, it searches the middle third.
+        (_, _, _) => {
+            *left = first_mid + 1;
+            *right = second_mid - 1;
+        }
     }
 
-    #[test]
-    fn returns_first_index_if_end_out_of_bounds() {
-        let index = ternary_search(&1, &[1, 2, 3], 0, 3);
-        assert_eq!(index, Some(0));
-    }
+    false
+}
 
-    #[test]
-    fn returns_last_index() {
-        let index = ternary_search(&3, &[1, 2, 3], 0, 2);
-        assert_eq!(index, Some(2));
-    }
+/// Determines if the given array is sorted in ascending order.
+///
+/// This helper function checks if the first element of the array is less than the
+/// last element, indicating an ascending order. It returns `false` if the array
+/// has fewer than two elements.
+///
+/// # Parameters
+///
+/// - `arr`: A slice of the array to check.
+///
+/// # Returns
+///
+/// A `bool` indicating whether the array is sorted in ascending order.
+fn is_asc_arr<T: Ord>(arr: &[T]) -> bool {
+    arr.len() > 1 && arr[0] < arr[arr.len() - 1]
+}
 
-    #[test]
-    fn returns_last_index_if_end_out_of_bounds() {
-        let index = ternary_search(&3, &[1, 2, 3], 0, 3);
-        assert_eq!(index, Some(2));
-    }
+#[cfg(test)]
+mod tests {
+    use super::*;
 
-    #[test]
-    fn returns_middle_index() {
-        let index = ternary_search(&2, &[1, 2, 3], 0, 2);
-        assert_eq!(index, Some(1));
+    macro_rules! test_cases {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (item, arr, expected) = $test_case;
+                    assert_eq!(ternary_search(&item, arr), expected);
```

Comment:
I think it will not hurt do add such check of the test data:
```suggestion
                    if let Some(expected_index) = expected {
                        assert_eq!(arr[expected_index], item);
                    }
                    assert_eq!(ternary_search(&item, arr), expected);
```